src/participant/participant.ts

```diff
@@ -1446,8 +1448,29 @@ export default class ParticipantController {
       log.info('Docs chatbot created for chatId', chatId);
     }
 
+    const history = Prompts.docs.getHistoryMessages({
+      connectionNames: this._getConnectionNames(),
+      context: context,
+    });
+
+    log.info(
+      'Sending history to chatbot',
+      history.map((message: vscode.LanguageModelChatMessage) =>
+        getContent(message)
+      )
+    );
+
+    const previousMessages =
+      history.length > 0
```

Comment:
So... I left it like this as a simple implementation but this needs so more discussion.

### Limitations

- There's no way to just add previous messages as context (at least not yet) through the app. We can make requests and wait for a full response for each previous message which is quite long.
- So, instead, we could include previous messages by making it all part of the same prompt and appending the last part of the request in the end.

### Risks

1.  Prompts end up unnecessarily too long and we easily reach the maximum prompt count.
2.  The earlier messages / prompts end up overtaking the question or are drastically different (i.e. user decided to ask something about docs which has nothing to do with what he was doing before and now the docs is referring to all that context and ends up less useful.

### Observations

- I think 1 is definitely a real risk especially if the user has been using the chat for a while and suddenly decides to use documentation.
- In my testing, I don't actually see much of 2 happening even when I am trying to be nefarious; seems like the end of the prompt does take a lot of precedence, so that's good.

### Mitigation Strategies

- We can deal with 1 and 2 by introducing a small cap on how many previous messages we include, limited to as low as 2-3?
- We could also make the prompt being sent more explained to the LLM i.e start with header "For context, previous prompts were: .... \n\n The prompt is: ....` or adding a `------` after the context. So far, I have found lack of clear separation not to be a problem.
- We can also request for the API to add some context-adding endpoint but this will probably involve a lot of waiting

Reply:
FYI the docs team is considering changing the way the endpoint works by not keeping history on the server but instead similarly as the copilot does, requiring to send the whole context with each request. But as you said, this won't, probably, happen soonish, so I like your idea to limit the history messages we include. About 3 sounds reasonable. I am also wondering if we want to include ChatResponseTurn, maybe it would be enough to just pass ChatRequestTurn?

Reply:
Let's try then with 2 questions & responses 👍